pandas/tests/reshape/test_pivot.py

```diff
@@ -1565,8 +1565,9 @@ def test_crosstab_normalize(self):
                               full_normal)
         tm.assert_frame_equal(pd.crosstab(df.a, df.b, normalize='index'),
                               row_normal)
-        tm.assert_frame_equal(pd.crosstab(df.a, df.b, normalize='columns'),
-                              col_normal)
+        tm.assert_frame_equal(
+            pd.crosstab(df.a, df.b, normalize='columns').astype('f8'),
```

Comment:
The change is happening because ATM the casting is done at the Block-level, while with the PR casting is done at the column-level.  `pd.crosstab(df.a, df.b, normalize='columns')` ATM returns

```
b    3    4
a          
1  0.5  0.0
2  0.5  1.0
```

and under the PR gives:

```
b    3    4
a          
1  0.5  0
2  0.5  1
```

i.e. the `4` column is `int64`.  Note that the original inputs `df.a` and `df.b` are both int64 to begin with.  I don't know `crosstab` well enough to comment on the ideal output, but it's hard to call this less-correct.

Reply:
I think the correct way to answer this is ask "was the upcast from int to float the result of coercing to a single dtype for multiple columns?" If so, then we should preserve integer dtype where we can, i.e.


For an all-integer `crosstab`, we preserve integer dtype:

```python
In [27]: a = pd.Series([1, 2])

In [28]: pd.crosstab(a, a)
Out[28]:
col_0  1  2
row_0
1      1  0
2      0  1
``` 

so this change seems good (and maybe worth explicitly testing rather than astyping the output, to ensure that we don't regress)?